nixos/modules/services/networking/syncthing.nix

```diff
@@ -51,25 +53,69 @@ let
-        "devices": ('${escapeShellArg (builtins.toJSON devices)}'${optionalString (cfg.settings.devices == {} || ! cfg.overrideDevices) " + .devices"}),
-        "folders": ('${escapeShellArg (builtins.toJSON folders)}'${optionalString (cfg.settings.folders == {} || ! cfg.overrideFolders) " + .folders"})
-    }''})
-
-    # send the new config
-    curl -X PUT -d "$new_cfg" ${cfg.guiAddress}/rest/config
-
+  '' +
+  /* Syncthing's rest API for the folders and devices is identical. Hence we
+  iterate the cfg.settings.{devices,folder} Nix variables using the
+  conf_type which is either "devices" or "folders": */
+  (lib.concatStringsSep "\n" (map (conf_type: let
+    new_conf_IDs = (concatStringsSep " " {
+      devices = devicesIDs;
+      folders = foldersIDs;
+    }.${conf_type});
+    override = {
+      devices = cfg.overrideDevices;
+      folders = cfg.overrideFolders;
+    }.${conf_type};
+    conf = {
+      inherit devices folders;
+    }.${conf_type};
+    # The API's id naming is slightly different for devices and folders
+    GET_IdAttrName = {
+      devices = "deviceID";
+      folders = "id";
+    }.${conf_type};
+    # All URLs and curl commands are based on: https://docs.syncthing.net/rest/config.html
+    baseAddress = "${cfg.guiAddress}/rest/config/${conf_type}";
+  in ((lib.optionalString (new_conf_IDs != []) ''
+  # We iterate the IDs list, and run a curl -X POST command for each, that
+  # should update that device/folder only.
+  for id in ${new_conf_IDs}; do
+    new_cfg=$(echo ${escapeShellArg (builtins.toJSON conf)} | ${jq} '.[] | select(.${GET_IdAttrName} == "'$id'")')
+    # Quoting https://docs.syncthing.net/rest/config.html: > PUT takes an
+    # array and POST a single object. In both cases if a given folder/device
+    # already exists, it’s replaced, otherwise a new one is added.
+    curl -d "$new_cfg" -X POST ${baseAddress}
+  done
+  '') +
+  /* If we need to override devices/folders, we iterate all currently configured
+  IDs, via another `curl -X GET`, and we delete all IDs that are not part of
+  the Nix configured list of IDs
+  */
+  (optionalString override ''
+    old_conf_${conf_type}_ids="$(curl -X GET ${baseAddress} | ${jq} --raw-output '.[].${GET_IdAttrName}')"
+    for id in ''${old_conf_${conf_type}_ids}; do
+      if echo ${new_conf_IDs} | grep -q $id; then
+        continue
+      else
+        curl -X DELETE ${baseAddress}/$id
+      fi
+    done
+  ''))) ["devices" "folders"])) +
+  /* Now we update the subOption "options", "gui" and "ldap" one by one, in
+  order to not override the folders and devices we handled in the previous
+  Nix Loop. */
+  (lib.concatStringsSep "\n" (map (subOption:
+    optionalString (builtins.hasAttr subOption cleanedConfig) ''
+      curl -X PUT -d ${escapeShellArg (builtins.toJSON cleanedConfig.${subOption})} \
+        ${cfg.guiAddress}/rest/config/${subOption}
+    '') [ "options" "gui" "ldap" ])
```

Comment:
wouldn't it be better to filter those keys we already pushed? maybe syncthing adds new config keys at some point and then we have to touch this code again

Reply:
> wouldn't it be better to filter those keys we already pushed? maybe syncthing adds new config keys at some point and then we have to touch this code again

Great idea! It also prevents `curl` commands from being concatenated if e.g the `subOption` is not defined in `cleanedConfig`.